### ISSUE TITLE
ORCA transforms LEFT JOIN into INNER JOIN in a query with WHERE condition

### ISSUE THREAD

**Issue by rovan** [greenplum-db/gpdb#4433]
### Greenplum version or build
```
PostgreSQL 8.3.23 (Greenplum Database 5.3.0 build dev) on x86_64-pc-linux-gnu, compiled by GCC gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16), 64-bit 
ORCA version
GPOPT version: 2.51.1, Xerces version: 3.1.2
```
### OS version and uname -a
```
CentOS Linux release 7.3.1611 (Core)
Linux 3.10.0-514.26.1.el7.x86_64 #1 SMP Thu Jun 29 16:05:25 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
autoconf options used ( config.status --config )
```
### autoconf options used ( config.status --config )
```
./config.status --config
'--prefix=/home/centos/gpbin' '--with-gssapi' '--with-libedit-preferred' '--with-python' '--with-perl' '--with-openssl' '--with-ldap' '--enable-pxf' '--with-libxml' '--enable-depend' '--enable-orca' '--enable-gpperfmon'
```

### Installation information ( pg_config )
```
BINDIR = /usr/local/greenplum-db-release/bin
DOCDIR = /usr/local/greenplum-db-release/doc/postgresql
HTMLDIR = /usr/local/greenplum-db-release/doc/postgresql
INCLUDEDIR = /usr/local/greenplum-db-release/include
PKGINCLUDEDIR = /usr/local/greenplum-db-release/include/postgresql
INCLUDEDIR-SERVER = /usr/local/greenplum-db-release/include/postgresql/server
LIBDIR = /usr/local/greenplum-db-release/lib
PKGLIBDIR = /usr/local/greenplum-db-release/lib/postgresql
LOCALEDIR = /usr/local/greenplum-db-release/share/locale
MANDIR = /usr/local/greenplum-db-release/share/man
SHAREDIR = /usr/local/greenplum-db-release/share/postgresql
SYSCONFDIR = /usr/local/greenplum-db-release/etc/postgresql
PGXS = /usr/local/greenplum-db-release/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/local/greenplum-db-release/include
CFLAGS = -O3 -std=gnu99 -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -I/usr/local/greenplum-db-release/include
CFLAGS_SL = -fpic
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/home/centos/gpbin/lib',--enable-new-dtags -L/usr/local/greenplum-db-release/lib
LDFLAGS_SL =
LIBS = -lpgport -lgpopt -lnaucrates -lgpdbcost -lgpos -lxerces-c -lxml2 -lrt -lgssapi_krb5 -lcrypt -ldl -lm -L/usr/local/greenplum-db-release/lib
VERSION = PostgreSQL 8.3.23
```
### Expected behavior

```
create table src as( select 1 as src_k, 't' as src_v union all select 2 as src_k, 'r' as src_v);
create table trg as( select 2 as trg_k, 'e' as trg_v union all select 3 as trg_k, 'y' as trg_v);

set optimizer=on;
select * from src left join trg on src_k = trg_k where src_v is distinct from trg_v

 src_k | src_v | trg_k | trg_v
-------+-------+-------+-------
     1 | t     |       |
     2 | r     |     2 | e
```

### Actual behavior
```
create table src as( select 1 as src_k, 't' as src_v union all select 2 as src_k, 'r' as src_v);
create table trg as( select 2 as trg_k, 'e' as trg_v union all select 3 as trg_k, 'y' as trg_v);

set optimizer=on;
select * from src left join trg on src_k = trg_k where src_v is distinct from trg_v

 src_k | src_v | trg_k | trg_v
-------+-------+-------+-------
     2 | r     |     2 | e
```
### Step to reproduce the behavior
```
create table src as( select 1 as src_k, 't' as src_v union all select 2 as src_k, 'r' as src_v);
create table trg as( select 2 as trg_k, 'e' as trg_v union all select 3 as trg_k, 'y' as trg_v);

select 'Normal behaviour' as example;

set optimizer=off;
select * from src left join trg on src_k = trg_k where src_v is distinct from trg_v;

select 'Erroneous behaviour' as example;

set optimizer=on;
select * from src left join trg on src_k = trg_k where src_v is distinct from trg_v;
```
Pivotal Greenplum 4.3.20.0 shows the same behavior as shown above.

**Comment by kapustor** [comment]
5.4.1 (ORCA  2.53.11) facing the same issue.

**Comment by 4etvegr** [comment]
Same on 4.3.20 with:
db_test=# select gp_opt_version();
                gp_opt_version
----------------------------------------------
 GPOPT version: 2.53.4, Xerces version: 3.1.1
(1 row)

**Comment by vraghavan78** [comment]
Thank you for reporting this. We will look into it today.

V
ᐧ

**Comment by hsyuan** [comment]
expression preprocessor did it wrong. We should not transform outer join to inner join for IDF predicate.

**Comment by vraghavan78** [comment]
Thanks Haisheng.

While we are fixing this, please set optimizer_enable_outerjoin_rewrite =
off; in your system.

We are actively working on it. So expect a fix very soon.

V
ᐧ

**Comment by khannaekta** [comment]
Fixed in ORCA version 2.54.4
